src/Web/Slack/Experimental/Views.hs

```diff
@@ -0,0 +1,139 @@
+import Web.Slack.Types (UserId)
+
+-- | View definition for some Slack surface. Has an inner type of either
+-- 'ModalView' or 'HomeTabView'.
+--
+-- <https://api.slack.com/reference/surfaces/views>
+data SlackView inner = SlackView
+  { blocks :: Vector B.SlackBlock
+  , privateMetadata :: Maybe Text
+  , callbackId :: Maybe Text
+  , externalId :: Maybe Text
+  , inner :: inner
+  }
+  deriving stock (Show, Eq, Generic)
+
+instance (FromJSON inner) => FromJSON (SlackView inner) where
+  parseJSON = withObject "SlackView" \o -> do
+    blocks <- o .: "blocks"
+    privateMetadata <- o .:? "private_metadata"
+    callbackId <- o .:? "callback_id"
+    externalId <- o .:? "external_id"
+    inner <- parseJSON @inner (A.Object o)
+    pure SlackView {..}
+
+instance (ToJSON inner) => ToJSON (SlackView inner) where
+  toJSON SlackView {..} = case toJSON (inner) of
+    A.Object innerObj ->
+      A.Object $
+        KM.fromList
+          [ "blocks" .= blocks
+          , "private_metadata" .= privateMetadata
+          , "callback_id" .= callbackId
+          , "external_id" .= externalId
+          ]
+          <> innerObj
+    _ -> error "inner of SlackView is not an object"
+
+data HomeTabView = HomeTabView
+  { type_ :: Expected "home"
+  }
+  deriving stock (Show, Eq, Generic)
+
+$(deriveJSON snakeCaseOptionsEatTrailing ''HomeTabView)
+
+-- | Modal view
+--
+-- <https://api.slack.com/reference/surfaces/views#modal>
+data ModalView = ModalView
+  { type_ :: Expected "modal"
+  , title :: SlackPlainTextOnly
+  -- ^ Title of the modal on the top left. Maximum length of 24 characters.
+  , close :: Maybe SlackPlainTextOnly
+  -- ^ Text appearing on the close button on the bottom-right of the modal.
+  -- Maximum length of 24 characters.
+  , submit :: Maybe SlackPlainTextOnly
+  -- ^ Text appearing on the submit buttonn. Maximum length of 24 characters.
+  -- Must be 'SlackPlainTextOnly'.
+  , submitDisabled :: Maybe Bool
+  -- ^ Whether one or more inputs must be filled before enabling the submit button.
+  -- For configuration modals: <https://api.slack.com/reference/workflows/configuration-view>
+  }
+  deriving stock (Show, Eq, Generic)
+
+$(deriveJSON snakeCaseOptionsEatTrailing ''ModalView)
+
+-- | Publishes the App Home view for a user.
+--
+-- <https://api.slack.com/methods/views.publish>
+data PublishReq = PublishReq
+  { userId :: UserId
+  -- ^ User to whom the view is being published.
+  , view :: HomeTabView
+  -- ^ View payload.
+  }
+  deriving stock (Show)
+
+instance ToForm PublishReq where
+  toForm PublishReq {..} =
+    [ ("user_id" :: Text, toQueryParam userId)
+    , ("view", toQueryParam . decodeUtf8 $ A.encode view)
+    ]
+
+data PublishResp = PublishResp
+  { view :: SlackView HomeTabView
+  }
+  deriving stock (Show)
+
+$(deriveJSON snakeCaseOptions ''PublishResp)
+
+type Api =
+  "views.publish"
+    :> AuthProtect "token"
+    :> ReqBody '[FormUrlEncoded] PublishReq
+    :> Post '[JSON] (ResponseJSON PublishResp)
```

Comment:
Non-blocking: I looked at [the docs for `:>`](https://hackage.haskell.org/package/servant-0.20.2/docs/Servant-API.html#t::-62-) and I'm still not sure I understand it.

I trust that this is correct, but might be worth a comment? Is like... each part here an element of a URL path? But `ReqBody` and `Post` are other parts of the HTTP request...